### DISCUSSION TITLE
@formily/reactive ArraySet 的 forEach 方法实现的不是很安全

### DISCUSSION THREAD

**Discussion by liuweiGL** [alibaba/formily#3225] [Q&A 2.x] [answered]
## 问题
`ArraySet` 的 value 属性在 forEach  遍历过程中有可能发生改变，当 len 大于 this.value.length  时传递给下游的值为 undefined 导致 `notifyObservers` 方法报错：
https://github.com/alibaba/formily/blob/297532f8c3b31bf4a85b7ef8c9579adb94f675b0/packages/reactive/src/reaction.ts#L95

![Snipaste_2022-06-24_17-20-24](https://user-images.githubusercontent.com/21048878/175509005-1ca2672b-973b-4303-a04d-4b307ba5818b.jpg)

注意看报错上方的打印日志，对应代码为：
```ts
  ArraySet2.prototype.forEach = function (callback) {
    if (this.value.length === 0) return
    for (var index = 0, len = this.value.length; index < len; index++) {
      callback(this.value[index])
      if (this.value.length !== len) {
        console.log(len, this.value)
      }
    }
  }
```

## 重现
我尝试根据报错写一个重现案例没成功，因为我这个页面使用了大量的自定义组件不晓得是哪块代码引发的问题。这也是没提 issue 的原因，不确定是框架层面的问题还是使用方式错误。

## 建议
这块是不是可以改成 `this.value.forEach` 或者 `for of` ？

https://github.com/alibaba/formily/blob/297532f8c3b31bf4a85b7ef8c9579adb94f675b0/packages/reactive/src/array.ts#L36

**Comment by janryWang** [comment]
是因为callback内部执行的时候有可能走到forEachDeleate导致长度发生变化是吗？

**Reply by liuweiGL** [reply]
我不确定是哪个地方导致的，可以肯定的是 forEach 过程中 value 的长度有可能会发生变化

**Comment by janryWang** [comment, marked as answer]
https://github.com/alibaba/formily/pull/3233 优化了下代码算法，核心处理就是对delete考虑并发问题
